plugin/route53/route53.go

```diff
@@ -15,98 +24,275 @@ import (
+	if err := h.updateZones(ctx); err != nil {
+		return err
 	}
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				log.Infof("breaking out of Route53 update loop: %v", ctx.Err())
+				return
+			case <-time.After(1 * time.Minute):
+				if err := h.updateZones(ctx); err != nil && ctx.Err() == nil /* Don't log error if ctx expired. */ {
+					log.Errorf("failed to update zones: %v", err)
+				}
+			}
+		}
+	}()
+	return nil
+}
 
-	if len(answers) == 0 {
-		return plugin.NextOrFailure(rr.Name(), rr.Next, ctx, w, r)
+// ServeDNS implements the plugin.Handler.ServeDNS.
+func (h *Route53) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	state := request.Request{W: w, Req: r}
+	qname := state.Name()
+
+	zName := plugin.Zones(h.zoneNames).Matches(qname)
+	if zName == "" {
+		return plugin.NextOrFailure(h.Name(), h.Next, ctx, w, r)
+	}
+	z, ok := h.zones[zName]
+	if !ok || z == nil {
+		return dns.RcodeServerFailure, nil
 	}
 
 	m := new(dns.Msg)
 	m.SetReply(r)
 	m.Authoritative, m.RecursionAvailable = true, true
-	m.Answer = answers
+	var result file.Result
+	h.zMu.RLock()
+	m.Answer, m.Ns, m.Extra, result = z.z.Lookup(state, qname)
+	h.zMu.RUnlock()
+
+	switch result {
+	case file.Success:
+	case file.NoData:
+	case file.NameError:
+		m.Rcode = dns.RcodeNameError
+	case file.Delegation:
+		m.Authoritative = false
+	case file.ServerFailure:
+		return dns.RcodeServerFailure, nil
+	}
 
 	state.SizeAndDo(m)
 	m, _ = state.Scrub(m)
 	w.WriteMsg(m)
 	return dns.RcodeSuccess, nil
 }
 
-func a(zone string, rrss []*route53.ResourceRecordSet) []dns.RR {
-	answers := []dns.RR{}
-	for _, rrs := range rrss {
-		for _, rr := range rrs.ResourceRecords {
-			r := new(dns.A)
-			r.Hdr = dns.RR_Header{Name: zone, Rrtype: dns.TypeA, Class: dns.ClassINET, Ttl: uint32(aws.Int64Value(rrs.TTL))}
-			r.A = net.ParseIP(aws.StringValue(rr.Value)).To4()
-			answers = append(answers, r)
+func awsStrTypeToUint16(t string) (uint16, bool) {
+	var out uint16
+	switch t {
+	case "A":
+		out = dns.TypeA
+	case "AAAA":
+		out = dns.TypeAAAA
+	case "CNAME":
+		out = dns.TypeCNAME
+	case "PTR":
+		out = dns.TypePTR
+	case "SOA":
+		out = dns.TypeSOA
+	case "NS":
+		out = dns.TypeNS
+	default:
+		return dns.TypeNone, false
+	}
+	return out, true
+}
+
+func parseSOA(soa string) (ns, mbox string, serial, refresh, retry, expire, minttl int, err error) {
+	parts := strings.SplitN(soa, " ", 7)
+	if len(parts) != 7 {
+		err = fmt.Errorf("failed to parse SOA record: %q", soa)
+		return
+	}
+	ns, mbox = parts[0], parts[1]
+	if serial, err = strconv.Atoi(parts[2]); err != nil {
+		return
+	}
+	if refresh, err = strconv.Atoi(parts[3]); err != nil {
+		return
+	}
+	if retry, err = strconv.Atoi(parts[4]); err != nil {
+		return
+	}
+	if expire, err = strconv.Atoi(parts[5]); err != nil {
+		return
+	}
+	if minttl, err = strconv.Atoi(parts[6]); err != nil {
+		return
+	}
+	return
+}
+
+func setRRValue(rr dns.RR, hdr *dns.RR_Header, value string) error {
```

Comment:
I'm not sure, but I feel there should be a better way to do this as well...

Reply:
ditto. done.